allow retrying initial ssh connection

dontlaugh opened the pull request.

Commit:
allow retrying initial ssh connection

dontlaugh committed:
update/correct BUILDING.md

dontlaugh commented:
@dcole-inviqa I needed a retry mechanism for my own infrastructure, too! I amended your commit with the recommendations from #1 

@bradrydzewski Let me know if there's anything else you'd like.

bradrydzewski commented on engine/engine_impl.go:
```diff
@@ -75,31 +76,18 @@ func (e *engine) Setup(ctx context.Context, spec *Spec) error {
-			Debug("failed to dial server")
 		return err
 	}
-	defer client.Close()
```
any reason this is being removed?  I believe we still need to close the client connection once we are done.

bradrydzewski replied:
this actually may be safer:

```
	client, err := e.retryDial(ctx, spec)
	if client != nil {
		defer client.Close()
	}
```

we use the above in our macstadium runner

dontlaugh replied:
This was probably a mistake. I'll update.

bradrydzewski commented:
we also have a more robust retry function here:
https://github.com/drone-runners/drone-runner-macstadium/blob/master/engine/engine.go#L373:L424

one important difference in the above function is that we close client connections if not nil before we retry. This prevents a potential fd leadk.

dontlaugh committed:
Code review feedback

* close ssh client explicitly before retrying
* retry based on a long network timeout instead of N attempts

dontlaugh commented:
Thanks for the review! I think I've addressed your comments.

I've followed your advice, but I'm curious about the file descriptor leak comment. It looks like it shouldn't happen from the code, as long as the underlying ssh library does what it's supposed to. Have you seen resources leak in the past? Does it show up on high-traffic servers?
